pkg/corpus/prio.go

```diff
@@ -19,7 +19,7 @@ type ProgramsList struct {
 	accPrios []int64
 }
 
-func (pl *ProgramsList) ChooseProgram(r *rand.Rand) *prog.Prog {
+func (pl *ProgramsList) Choose(r *rand.Rand) *prog.Prog {
```

Comment:
You mentioned some other priority selection algorithm (choose 2 random, then choose the one with high priority). Will it work here? It can make it easier to have priority that depends on the total number of programs (won't need to recalculate prio for each program after addition of each new program).

Reply:
> choose 2 random, then choose the one with high priority

This will only approximate the current algorithm used in `prio.go` -- weighting by the prog's `|signal|`. And only if the distribution of `|signal|` is not skewed.

> It would be nice to localize all of the priority logic here w/o leaking it to pkg/fuzzer. 

I still cannot come up with a good formula that also covers well the possible corner cases:
1. There are only a few corpus programs that we want to focus on.
2. There are many corpus programs we want to focus on.

These cases are quite real and we should be ready to handle them.

If the delta between high and low prio is small, we won't really focus the fuzzer on anything. If it's too high, we will never fuzz non-prioritized seeds. Not sure there even exists an approach that assigns static prios and covers all cases.

In any case, I can move the already implemented logic to `prio.go`. I guess, in `corpus.go` it will be something like 
```
type Config struct {
  // ...
  Priorities []PriorityGroup
}

type PriorityGroup struct {
  Name string
  PCs  map[uint64]struct{}
}
```

Then `pkg/fuzzer` will not be taking care about it.

Reply:
1. Let's localize it all in prio.go as much as possible. This looks positive regardless of the rest.

2. Can we combine configuration of this with existing coverage filter? They are effectively doing the same. It's even unclear what's the priority (what if something is in coverage filter, but not in new groups, or the other way around?).
The old coverage filter is effectively a group with priority 1, and the rest (what's not listed) is priority 0.

3. Instead of names it would be more useful to use weight (say float 0..1 that says how frequently this group is selected). If I am every using this in config, how I will come up with names for them. Names are also effectively not used by the code.

4. It would be useful to have fallback group/priority for "everything else" (what's not listed explicitly). Then it will be possible to say "concentrate on these parts, but test the whole kernel).

Reply:
Added these two commits:

* [pkg/corpus: support multiple focus areas](https://github.com/google/syzkaller/pull/5415/commits/5f0c94d6937f8ba257f319039826c41831c7fad7) 
* [pkg/mgrconfig, syz-manager: support focus areas](https://github.com/google/syzkaller/pull/5415/commits/6199fbcd81e4e515252c14731313a033f98fa4e8) 

PTAL